File: review/palindrome-partitioning.ts
/**
 * https://leetcode.com/problems/palindrome-partitioning/description/
 */
export { partition };

function partition(s: string): string[][] {
  const result: string[][] = [];

  function backtrack(start: number, path: string[]) {
    if (start >= s.length) {
      result.push([...path]);
      return;
    }

    for (let end = start + 1; end <= s.length; end++) {
      const evaluate = s.substring(start, end);
      if (isPalindrome(evaluate)) {
        path.push(evaluate);
        backtrack(end, path);
        path.pop();
      }
    }
  }

  backtrack(0, []);
  return result;
}

function isPalindrome(s: string): boolean {
  let l = 0;
  let r = s.length - 1;
  while (l < r) {
    if (s[l] !== s[r]) return false;
    l++;
    r--;
  }
  return true;
}

File: python/today10/palindrome_partitioning.py
from typing import List

class Solution:
    def partition(self, s: str) -> List[List[str]]:
        result = []
        part = []
        def backtrack(i: int) -> None:
            if i >= len(s):
                result.append(part.copy())
                return
            
            for j in range(i,len(s)):
                if self.isPalindrome(s,i,j):
                    part.append(s[i:j+1])
                    backtrack(j+1)
                    part.pop()
        backtrack(0)
        return result
    
    def isPalindrome(self, s: str, i: int, j: int) -> bool:
        while i < j:
            if s[i] != s[j]:
                return False
            i, j = i+1, j-1
        return True